### PR TITLE
fix: reduce lock contention in `RepartitionExec::execute`

### PR DESCRIPTION
**Pull request by crepererum**
## Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/10014

## Rationale for this change
The state is initialized ONCE for all partitions. However this may take a short while (on a very busy system 1ms or more). It is quite likely that multiple threads call `execute` at the same time, because we have just fanned out to the number "target partitions" which is likely set to the number of CPU cores which now all try to start to execute the plan at the same time.

The solution is to not waste CPU circles in some futex lock but to tell the async runtime (= tokio) that we are performing work and the other threads should rather do something useful.

## What changes are included in this PR?
This mostly just moves code around, no functional change intended.

## Are these changes tested?
Existing tests still pass.

## Are there any user-facing changes?
Faster query exec.

### CODE REVIEW

File: datafusion/physical-plan/src/repartition/mod.rs [hunk 1]
```diff
@@ -77,6 +79,90 @@ struct RepartitionExecState {
...
+        let mut spawned_tasks = Vec::with_capacity(num_input_partitions);
+        for i in 0..num_input_partitions {
+            let txs: HashMap<_, _> = channels
+                .iter()
+                .map(|(partition, (tx, _rx, reservation))| {
+                    (*partition, (tx[i].clone(), Arc::clone(reservation)))
+                })
+                .collect();
+
+            // TODO: metric input-output mapping is broken
```
Comment (crepererum) [inline datafusion/physical-plan/src/repartition/mod.rs:134]: This was broken before: the first parameter was set to `partition`, i.e. the first partition that initializes the state. That's clearly wrong. I now initialize it to `0` and will fix the tracking properly in a follow-up PR.

File: datafusion/physical-plan/src/repartition/mod.rs [hunk 2]
```diff
@@ -1240,7 +1294,10 @@ mod tests {
         std::mem::drop(output_stream0);
 
         // Now, start sending input
-        input.wait().await;
+        let mut background_task = JoinSet::new();
```
Comment (alamb) [inline datafusion/physical-plan/src/repartition/mod.rs:1297]: What is the purpose of this change? I tried this change without the other changes in this PR and the test still passes (I was expecting it would hang or something)

File: datafusion/physical-plan/src/repartition/mod.rs [hunk 3]
```diff
@@ -77,6 +79,90 @@ struct RepartitionExecState {
...
+        }
+
+        Self {
+            channels,
+            abort_helper: Arc::new(spawned_tasks),
+        }
+    }
+}
+
+type LazyState = Arc<OnceCell<Mutex<RepartitionExecState>>>;
```
Comment (alamb) [inline datafusion/physical-plan/src/repartition/mod.rs:164]: I think including the context from your PR description about *why* this is using a (`tokio`) `OnceCell` would help avoid future regressions (if someone in the future concluded incorrectly, for example, that the overhead of `async::OnceCell` was not needed)

```suggestion
//// Lazily initialized state
////
/// Note that the state is initialized ONCE for all partitions by a single task(thread).
/// This may take a short while.  It is also like that multiple threads 
/// call execute at the same time, because we have just started "target partitions" tasks
/// which is commonly set to the number of CPU cores and all call execute at the same time.
///
/// Thus, use a **tokio** `OnceCell` for this initialization so as not to waste CPU cycles 
/// in a futex lock but instead allow other threads to do something useful.
///
/// Uses a parking_lot `Mutex` to control other accesses as they are very short duration
///  (e.g. removing channels on completion) where the overhead of `await` is not warranted. 
type LazyState = Arc<tokio::sync::OnceCell<Mutex<RepartitionExecState>>>;
```

File: datafusion/physical-plan/src/repartition/mod.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-plan/src/repartition/mod.rs:134]: Filed https://github.com/apache/arrow-datafusion/issues/10015 to track

**Review by alamb — APPROVED** [review]
Thank you @crepererum  -- I reviewed this code carefully and it makes sense to me

I think it is important to document the rationale in comments to reduce the chance of this change being accidentally undone in the future, but otherwise I think it is ready to go 🚀

**Comment by crepererum** [comment]
TBH a slightly slower runtime in a "repeat a single query in a hot-loop" benchmark is somewhat expected. If you have enough tokio threads to spare, you can totally futex-wait. But for any kind of server software that's a bad idea.

File: datafusion/physical-plan/src/repartition/mod.rs (hunk 2 above)
Comment (crepererum) [reply datafusion/physical-plan/src/repartition/mod.rs:1297]: https://github.com/apache/arrow-datafusion/pull/10009#discussion_r1557396828
